### PR TITLE
[kuberay] Test Ray client and update autoscaler image

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR adds KubeRay e2e testing for Ray client and updates the suggested autoscaler image to one running the merge commit of PR https://github.com/ray-project/ray/pull/23883 .

In more detail, the PR does these things:
- Adds Ray client and dashboard ports to the head service in the example config
- Adds basic testing for Ray client, by modifying the GPU test from [#23883](https://github.com/ray-project/ray/pull/23883) to submit tasks with Ray client
- Updates the autoscaler image to the one from this PR
- Documents context on why we're currently pinning this image
- Allows specifying the autoscaler image in the e2e test
  * I used this to confirm that the pinned autoscaler image is compatible with all Ray images since Ray 1.11.0 -- Ray 1.11.0, Ray 1.11.1 (soon to be released), Ray 1.12.0, and Ray 1.12.1 (soon to be released).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] e2e tests

### CODE REVIEW

File: python/ray/tests/kuberay/test_autoscaling_e2e.py [hunk 1]
```diff
@@ -254,12 +262,11 @@ def testAutoscaling(self):
...
         logger.info("Scheduling an Actor with GPU demands.")
-        kubectl_exec_python_script(
-            script_name="gpu_actor_placement.py",
-            pod=head_pod,
-            container="ray-head",
-            namespace="default",
-        )
+        # Use Ray client to validate that it works against KubeRay.
+        with ray_client_port_forward(
+            head_service="raycluster-complete-head-svc", ray_namespace="gpu-test"
```
Comment (Jeffwan) [inline python/ray/tests/kuberay/test_autoscaling_e2e.py:267]: What's the port it's forwarding? Should it be part of arguments?

Comment (Jeffwan) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py:267]: never mind, I find that client_port 10001 is the default arg in `ray_client_port_forward` definition.

**Comment by Jeffwan** [comment]
The changes looks good to me!

**Review by Jeffwan — APPROVED** [review]

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 2]
```diff
@@ -103,8 +112,12 @@ spec:
...
         # The Ray autoscaler sidecar to the head pod
         - name: autoscaler
-          # TODO: Use released Ray version starting with Ray 1.12.0.
-          image: rayproject/ray:413fe0
+          # The autoscaler image used carries the latest improvements to KubeRay autoscaling
+          # support.
+          # It is confirmed (via kuberay/test_autoscaling_e2e.py) to be compatible with all
+          # Ray versions since Ray 1.11.0.
+          # TODO: Use released Ray version when autoscaling support is stable.
+          image: rayproject/ray:8c5fe4
```
[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:120]: Yep, it's the GPU typo fix.

File: python/ray/tests/kuberay/test_autoscaling_e2e.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py:267]: 10001 is the default port for Ray client, so it seemed to make sense as a default set for this function.

**Comment by DmitriGekhtman** [comment]
> The change looks good to me!

Thanks for reviewing!

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -396,9 +396,9 @@
     - echo "--- Building py37-cpu Ray image for the test."
     - LINUX_WHEELS=1 ./ci/ci.sh build
     - pip install -q docker
-    - python ci/build/build-docker-images.py --py-versions py37 --device-types cpu --build-type LOCAL --build-base
+    - python ci/build/build-docker-images.py --py-versions py36 --device-types cpu --build-type LOCAL --build-base
```
Comment (pcmoritz) [inline .buildkite/pipeline.yml:399]: Was this on purpose? The comment in the echo statement above still says py37 (which probably makes more sense given that py36 is already EOL)

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 4]
```diff
@@ -8,13 +8,14 @@ metadata:
...
 spec:
-  rayVersion: '1.11.0'
+  rayVersion: '1.12.0'
   enableInTreeAutoscaling: false
   ######################headGroupSpecs#################################
   # head group template and specs, (perhaps 'group' is not needed in the name)
   headGroupSpec:
     # Kubernetes Service Type, valid values are 'ClusterIP', 'NodePort' and 'LoadBalancer'
     serviceType: ClusterIP
+    #enableIngress: true
```
Comment (pcmoritz) [inline python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:18]: Let's put a space after `#` to be consistent with the rest of the file

Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:18]: I left this line in accidentally -- will remove.

File: .buildkite/pipeline.yml (hunk 3 above)
Comment (DmitriGekhtman) [reply .buildkite/pipeline.yml:399]: > which probably makes more sense given that py36 is already EOL

Ah, thanks for the context.

The CI environment currently uses py36, so I changed the the image to py36 for Ray client compatibility.

A better move would probably be to upgrade the CI environment for this job to py37 instead. I will look into doing that.

Comment (DmitriGekhtman) [reply .buildkite/pipeline.yml:399]: Nice catch on the echo line -- we should keep that consistent.

Comment (DmitriGekhtman) [reply .buildkite/pipeline.yml:399]: cc @simon-mo for thoughts/context on our use of Python 3.6 in the Ray CI.

File: python/ray/tests/kuberay/utils.py [hunk 5]
```diff
@@ -227,3 +230,102 @@ def get_raycluster(raycluster: str, namespace: str) -> Dict[str, Any]:
...
+            "-n",
+            namespace,
+            "port-forward",
+            f"service/{service}",
+            f"{local_port}:{service_port}",
+        ]
+    )
+
+    def terminate_process():
+        process.terminate()
```
Comment (pcmoritz) [inline python/ray/tests/kuberay/utils.py:300]: Will this give an error if terminated twice? If yes, we might want to protect against it, so there are no ugly error messages on shutdown :)

Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/utils.py:300]: Hey, great point.
I'm pretty sure `process.terminate()`, but let me confirm and add a comment.

File: python/ray/tests/kuberay/test_autoscaling_e2e.py [hunk 6]
```diff
@@ -138,6 +122,30 @@ def _get_ray_cr_config(
         gpu_group["groupName"] = "fake-gpu-group"
         config["spec"]["workerGroupSpecs"].append(gpu_group)
 
+        # Substitute images.
+        for group_spec in config["spec"]["workerGroupSpecs"] + [
+            config["spec"]["headGroupSpec"]
+        ]:
+            containers = group_spec["template"]["spec"]["containers"]
+
+            # Will assume the example config maintains this invariant:
```
Comment (pcmoritz) [inline python/ray/tests/kuberay/test_autoscaling_e2e.py:131]: uh, this is dangerous -- you should add an `assert ray_container["name"] == "ray-head"` or similar here

File: python/ray/tests/kuberay/test_autoscaling_e2e.py [hunk 7]
```diff
@@ -138,6 +122,30 @@ def _get_ray_cr_config(
...
+
+            ray_container["image"] = RAY_IMAGE
+
+            for container in containers:
+                container["imagePullPolicy"] = PULL_POLICY
+
+        head_containers = config["spec"]["headGroupSpec"]["template"]["spec"][
+            "containers"
+        ]
+        autoscaler_container = [
```
Comment (pcmoritz) [inline python/ray/tests/kuberay/test_autoscaling_e2e.py:142]: Very interesting idiom!

**Review by pcmoritz — APPROVED** [review]

File: python/ray/tests/kuberay/test_autoscaling_e2e.py (hunk 6 above)
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py:131]: Will do -- will also rename the worker containers "ray-worker" from what is now (I think it's "ml" or something)

File: python/ray/tests/kuberay/test_autoscaling_e2e.py (hunk 7 above)
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py:142]: a for loop would have worked just as well... main function of a math phd is to induce weird programming habits

File: .buildkite/pipeline.yml (hunk 3 above)
Comment (DmitriGekhtman) [reply .buildkite/pipeline.yml]: Fixed to use Py37 in the test image and in the build step's host CI environment.

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml (hunk 4 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml]: Deleted stray line.

File: python/ray/tests/kuberay/utils.py (hunk 5 above)
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/utils.py:300]: Made the cleanup extra carefuly and commented that running it multiple times is unproblematic.

File: python/ray/tests/kuberay/test_autoscaling_e2e.py (hunk 6 above)
Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py]: Ray container for workers renamed "ray-worker". Test now asserts that the first container is one of `["ray-worker", "ray-head"]`.

**Comment by DmitriGekhtman** [comment]
Operator test passes. Test failures unrelated.

**Pull request merged**